File: frontend/src/api/endpoints/plans.ts
/**
 * Plan API functions.
 *
 * Backend source: app/routers/plans.py
 * Endpoints consumed:
 *   POST /generate-plan       → PlanJobResponse  (HTTP 202 Accepted)
 *   GET  /plan/{plan_id}      → BlockPlanResponse
 *
 * Notes on async generation flow:
 *   POST /generate-plan enqueues a Celery job and returns immediately with
 *   HTTP 202. The response contains a plan_id and a job_id. Callers must
 *   poll GET /plan/{plan_id} until plan.status transitions from
 *   "pending" / "generating" to "ready" or "failed".
 *   The polling logic lives at the component/hook layer, not here.
 *
 * Rules:
 *   - All functions go through the shared apiClient.
 *   - No mock data, no fallbacks, no invented response fields.
 */

import { apiClient } from '../client';
import type {
  BlockPlanResponse,
  PlanGenerateRequest,
  PlanJobResponse,
} from '../types';

// ---------------------------------------------------------------------------
// POST /generate-plan  (HTTP 202 Accepted)
// ---------------------------------------------------------------------------

/**
 * Enqueue an asynchronous block-plan generation job.
 *
 * The backend returns HTTP 202 immediately with a PlanJobResponse containing
 * the job handle. Poll GET /plan/{plan_id} for the completed result.
 *
 * @param request – horizon type/dates and optional corridor/department filter
 * @returns PlanJobResponse with accepted, job_id, plan_id, status
 */
export async function generatePlan(
  request: PlanGenerateRequest,
): Promise<PlanJobResponse> {
  const response = await apiClient.post<PlanJobResponse>(
    '/generate-plan',
    request,
  );
  return response.data;
}

// ---------------------------------------------------------------------------
// GET /plan/{plan_id}
// ---------------------------------------------------------------------------

/**
 * Retrieve a block plan by its ID.
 *
 * @param planId – positive integer plan identifier
 * @returns BlockPlanResponse including assignments and embedded KpiResponse
 * @throws ApiError with code "not_found" if the plan does not exist
 */
export async function getPlan(planId: number): Promise<BlockPlanResponse> {
  const response = await apiClient.get<BlockPlanResponse>(`/plan/${planId}`);
  return response.data;
}

File: backend/app/routers/plans.py
"""Block-plan HTTP routes. Business rules live in ``PlanService``."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends

from app.routers.responses import NOT_FOUND_RESPONSES, UNAVAILABLE_RESPONSES
from app.schemas.horizon import PlanGenerateRequest
from app.schemas.plans import BlockPlanResponse, PlanJobResponse
from app.services import get_plan_service
from app.services.plans import PlanService

router = APIRouter(tags=["plans"])


@router.post(
    "/generate-plan",
    response_model=PlanJobResponse,
    status_code=202,
    responses=UNAVAILABLE_RESPONSES,
    summary="Request asynchronous block-plan generation",
    description=(
        "Accepts a planning horizon and optional corridor/department filter. "
        "Enqueues background optimization job in Celery worker and returns "
        "a job handle immediately (HTTP 202 Accepted). "
        "Poll GET /plan/{plan_id} for the completed plan status."
    ),
)
def generate_plan(
    payload: PlanGenerateRequest,
    service: Annotated[PlanService, Depends(get_plan_service)],
) -> PlanJobResponse:
    return service.generate_plan(payload)


@router.get(
    "/plan/{plan_id}",
    response_model=BlockPlanResponse,
    responses=NOT_FOUND_RESPONSES,
    summary="Retrieve a completed block plan by ID",
)
def get_plan(
    plan_id: int,
    service: Annotated[PlanService, Depends(get_plan_service)],
) -> BlockPlanResponse:
    if plan_id < 1:
        from fastapi import HTTPException

        raise HTTPException(status_code=422, detail="plan_id must be greater than 0")
    return service.get_plan(plan_id)
